// hash-it/lib/constants.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

/**
 * @constant {Array<string>} OBJECT_CLASSES
 */
var OBJECT_CLASSES = exports.OBJECT_CLASSES = ['Arguments', 'Array', 'ArrayBuffer', 'Boolean', 'DataView', 'Date', 'Error', 'Float32Array', 'Float64Array', 'Function', 'Generator', 'GeneratorFunction', 'HTMLElement', 'Int8Array', 'Int16Array', 'Int32Array', 'Map', 'Math', 'Null', 'Number', 'Object', 'Promise', 'RegExp', 'Set', 'String', 'Symbol', 'Uint8Array', 'Uint8ClampedArray', 'Uint16Array', 'Uint32Array', 'Undefined', 'WeakMap', 'WeakSet'];

/**
 * @constant {Object} OBJECT_CLASS_MAP
 */
var OBJECT_CLASS_MAP = exports.OBJECT_CLASS_MAP = OBJECT_CLASSES.reduce(function (objectClasses, type) {
  objectClasses['[object ' + type + ']'] = type;

  return objectClasses;
}, {});

/**
 * @constant {Object} OBJECT_CLASS_TYPE_MAP
 */
var OBJECT_CLASS_TYPE_MAP = exports.OBJECT_CLASS_TYPE_MAP = Object.keys(OBJECT_CLASS_MAP).reduce(function (objectClassTypes, objectClass) {
  objectClassTypes[OBJECT_CLASS_MAP[objectClass].toUpperCase()] = objectClass;

  return objectClassTypes;
}, {});

/**
 * @constant {number} RECURSIVE_COUNTER_CUTOFF
 */
var RECURSIVE_COUNTER_CUTOFF = exports.RECURSIVE_COUNTER_CUTOFF = 512;

/**
 * @constant {Array<string>} REPLACE_RECURSIVE_VALUE_CLASSES
 */
var REPLACE_RECURSIVE_VALUE_CLASSES = exports.REPLACE_RECURSIVE_VALUE_CLASSES = [OBJECT_CLASS_TYPE_MAP.ARRAY, OBJECT_CLASS_TYPE_MAP.OBJECT];

/**
 * @constant {Array<string>} REPLACE_STRINGIFICATION_CLASSES
 */
var REPLACE_STRINGIFICATION_CLASSES = exports.REPLACE_STRINGIFICATION_CLASSES = [OBJECT_CLASS_TYPE_MAP.DATE, OBJECT_CLASS_TYPE_MAP.MAP, OBJECT_CLASS_TYPE_MAP.SET, OBJECT_CLASS_TYPE_MAP.REGEXP, OBJECT_CLASS_TYPE_MAP.ERROR, OBJECT_CLASS_TYPE_MAP.GENERATORFUNCTION, OBJECT_CLASS_TYPE_MAP.MATH, OBJECT_CLASS_TYPE_MAP.ARRAYBUFFER, OBJECT_CLASS_TYPE_MAP.DATAVIEW, OBJECT_CLASS_TYPE_MAP.FLOAT32ARRAY, OBJECT_CLASS_TYPE_MAP.FLOAT64ARRAY, OBJECT_CLASS_TYPE_MAP.INT8ARRAY, OBJECT_CLASS_TYPE_MAP.INT16ARRAY, OBJECT_CLASS_TYPE_MAP.INT32ARRAY, OBJECT_CLASS_TYPE_MAP.UINT8ARRAY, OBJECT_CLASS_TYPE_MAP.UINT8CLAMPEDARRAY, OBJECT_CLASS_TYPE_MAP.UINT16ARRAY, OBJECT_CLASS_TYPE_MAP.UINT32ARRAY, OBJECT_CLASS_TYPE_MAP.PROMISE, OBJECT_CLASS_TYPE_MAP.GENERATOR, OBJECT_CLASS_TYPE_MAP.WEAKMAP, OBJECT_CLASS_TYPE_MAP.WEAKSET];

/**
 * @constant {Array<string>} STRINGIFY_SELF_CLASSES
 */
var STRINGIFY_SELF_CLASSES = exports.STRINGIFY_SELF_CLASSES = [OBJECT_CLASS_TYPE_MAP.ARRAY, OBJECT_CLASS_TYPE_MAP.OBJECT, OBJECT_CLASS_TYPE_MAP.ARGUMENTS];

/**
 * @constant {Array<string>} STRINGIFY_PREFIX_CLASSES
 */
var STRINGIFY_PREFIX_CLASSES = exports.STRINGIFY_PREFIX_CLASSES = [OBJECT_CLASS_TYPE_MAP.ERROR, OBJECT_CLASS_TYPE_MAP.REGEXP];

/**
 * @constant {Array<string>} STRINGIFY_ITERABLE_CLASSES
 */
var STRINGIFY_ITERABLE_CLASSES = exports.STRINGIFY_ITERABLE_CLASSES = [OBJECT_CLASS_TYPE_MAP.MAP, OBJECT_CLASS_TYPE_MAP.SET];

/**
 * @constant {Array<string>} STRINGIFY_NOT_ENUMERABLE_CLASSES
 */
var STRINGIFY_NOT_ENUMERABLE_CLASSES = exports.STRINGIFY_NOT_ENUMERABLE_CLASSES = [OBJECT_CLASS_TYPE_MAP.PROMISE, OBJECT_CLASS_TYPE_MAP.GENERATOR, OBJECT_CLASS_TYPE_MAP.WEAKMAP, OBJECT_CLASS_TYPE_MAP.WEAKSET];

/**
 * @constant {Array<string>} STRINGIFY_PREFIX_JOIN_CLASSES
 */
var STRINGIFY_PREFIX_JOIN_CLASSES = exports.STRINGIFY_PREFIX_JOIN_CLASSES = [OBJECT_CLASS_TYPE_MAP.FLOAT32ARRAY, OBJECT_CLASS_TYPE_MAP.FLOAT64ARRAY, OBJECT_CLASS_TYPE_MAP.INT8ARRAY, OBJECT_CLASS_TYPE_MAP.INT16ARRAY, OBJECT_CLASS_TYPE_MAP.INT32ARRAY, OBJECT_CLASS_TYPE_MAP.UINT8ARRAY, OBJECT_CLASS_TYPE_MAP.UINT8CLAMPEDARRAY, OBJECT_CLASS_TYPE_MAP.UINT16ARRAY, OBJECT_CLASS_TYPE_MAP.UINT32ARRAY];

/**
 * @constant {Array<string>} STRINGIFY_SELF_TYPES
 */
var STRINGIFY_SELF_TYPES = exports.STRINGIFY_SELF_TYPES = ['string', 'number'];

/**
 * @constant {Array<string>} STRINGIFY_PREFIX_TYPES
 */
var STRINGIFY_PREFIX_TYPES = exports.STRINGIFY_PREFIX_TYPES = ['boolean', 'undefined', 'function', 'symbol'];

/**
 * @constant {Array<string>} STRINGIFY_TOSTRING_TYPES
 */
var STRINGIFY_TOSTRING_TYPES = exports.STRINGIFY_TOSTRING_TYPES = ['symbol', 'function'];

/**
 * @constant {Array<string>} STRINGIFY_TYPEOF_TYPES
 */
var STRINGIFY_TYPEOF_TYPES = exports.STRINGIFY_TYPEOF_TYPES = [].concat(STRINGIFY_SELF_TYPES, STRINGIFY_PREFIX_TYPES);

/**
 * @constant {RegExp} HTML_ELEMENT_REGEXP
 */
var HTML_ELEMENT_REGEXP = exports.HTML_ELEMENT_REGEXP = /\[object (HTML(.*)Element)\]/;

/**
 * @constant {Object} MATH_OBJECT
 */
var MATH_OBJECT = exports.MATH_OBJECT = ['E', 'LN2', 'LN10', 'LOG2E', 'LOG10E', 'PI', 'SQRT1_2', 'SQRT2'].reduce(function (mathObject, property) {
  var _extends2;

  return _extends({}, mathObject, (_extends2 = {}, _extends2[property] = Math[property], _extends2));
}, {});
